### PR TITLE
First functional implementation

### PR DESCRIPTION
**Pull request by LoyVanBeek**
Depends on:
- https://github.com/tue-robotics/yapykaldi/pull/1/
- https://github.com/tue-robotics/kaldi.git 

To test this:
- Either [setup audio loopback](https://unix.stackexchange.com/questions/82259/how-to-pipe-audio-output-to-mic-input) or speak at the appropriate time
- `roslaunch yapykaldi_ros yapykaldi_hmi.launch`
- `rosrun hmi test_query --grammar-file=orders.fcfg O`  # See attached file, from https://github.com/tue-robotics/grammar_parser/blob/master/README.md
- `aplay output.wav` (see attached file, me saying "Can I have  coke?")

[ordering.zip](https://github.com/tue-robotics/yapykaldi_ros/files/4821417/ordering.zip)

TODO:

- [x] Recording still is not working OK, recording a 2nd time crashes :cry: Mostly due to my botched implementation of the `PyaudioMicrophoneSource` in `yapykaldi`
- [ ] Voice Activity detection to stop listening at the right time

### CODE REVIEW

**Comment by LoyVanBeek** [comment]
Instead of an audio-based Voice Activity Detector, I simply use a timeout now. Nothing new heard for a second = user stopped speaking.

File: src/yapykaldi_ros/hmi_server_yapykaldi.py [hunk 1]
```diff
@@ -1,16 +1,92 @@
...
+        self._source.open()
+
+        rospy.loginfo("Setting up ASR, may take a while...")
+        self._asr = Asr(model_dir=os.path.expanduser(rospy.get_param('~model_dir')),
+                        model_type=rospy.get_param('~model_type', 'nnet3'),
+                        stream=self._source)
+        rospy.loginfo("Set up ASR")
+
+        self._asr.register_fully_recognized_callback(self.string_fully_recognized_callback)
+        self._asr.register_partially_recognized_callback(self.string_partially_recognized_callback)
```
Comment (ar13pit) [inline src/yapykaldi_ros/hmi_server_yapykaldi.py]:
```suggestion
        self._asr.register_callback(self.string_fully_recognized_callback)
        self._asr.register_callback(self.string_partially_recognized_callback, partial=True)
```

File: src/yapykaldi_ros/hmi_server_yapykaldi.py [hunk 2]
```diff
@@ -1,16 +1,92 @@
...
-import rospy
-from yapykaldi import Asr
 from hmi import AbstractHMIServer, HMIResult
+from hmi.common import parse_sentence, verify_grammar
+import os
+import rospy
+from threading import Event
+from yapykaldi.asr import Asr
+from yapykaldi.audio_handling.sources import WaveFileSource, PyAudioMicrophoneSource
+from yapykaldi.audio_handling.sinks import WaveFileSink
```
Comment (ar13pit) [inline src/yapykaldi_ros/hmi_server_yapykaldi.py]:
```suggestion
from yapykaldi.io import WaveFileSource, PyAudioMicrophoneSource, WaveFileSink
```

File: src/yapykaldi_ros/hmi_server_yapykaldi.py [hunk 3]
```diff
@@ -14,7 +15,7 @@ class HMIServerYapykaldi(AbstractHMIServer):
     """HMI server wrapper yapykaldi ASR app"""
     def __init__(self):
 
-        # route_logger_to_ros('yapykaldi')
+        # route_logger_to_ros('YAPYKALDI_LOGGER_NAME')
```
Comment (LoyVanBeek) [inline src/yapykaldi_ros/hmi_server_yapykaldi.py]: Extraneous quoting!

Comment (ar13pit) [reply src/yapykaldi_ros/hmi_server_yapykaldi.py]: Oops! My bad. Fixed.

**Comment by ar13pit** [comment]
I will merge this only after tue-robotics/yapykaldi#4 is closed as that will result in some changes in ASR API

**Review by ar13pit — APPROVED** [review]

**Pull request merged**